### PR TITLE
feat: Metrics Support in `tritonfrontend`

### PR DESCRIPTION
**Pull request by KrishnanPrash**
#### What does the PR do?
Adding support for `Metrics` in `tritonfrontend`.  This involves two components: 
- In tritonfrontend_pybind.cc, added bindings for `HTTPMetricsServer` 
- In tritonfrontend/_api/_metrics.py, added a `Metrics` class 

With this PR, similar to `KServeHttp` and `KServeGrpc`, the metrics service can used with:
```python
import tritonserver
from tritonfrontend import Metrics
server = tritonserver.Server(model_repostory=...).start(wait_until_ready=True)
metrics_service = Metrics(server)
metrics_service.start()
...
metrics_service.stop()
```

Additional Changes made in this PR:
- Modified test functions documentation based on this [comment](https://github.com/triton-inference-server/server/pull/7671#discussion_r1788161435)
- Removed extra parameter in `request.post(...)` based on this [comment](https://github.com/triton-inference-server/server/pull/7671#discussion_r1788429131)

#### Test plan:
Added 3 test function to L0_python_api:
- `test_metrics_default_port()` : Tests whether the metrics service can start as expected
- `test_metrics_custom_port()`: Tests whether arguments defined in `tritonfrontend.Metrics.Options` are passed successfully to `HTTPMetrics`
- `test_metrics_update()`: Tests whether `nv_inference_count` value goes from 0 to 1 if inference request is performed. 
- CI Pipeline ID: 19197748

### CODE REVIEW

File: qa/L0_python_api/test_model_repository/identity/config.pbtxt [hunk 1]
```diff
@@ -35,10 +35,18 @@ input [
...
     dims: [ 1 ]
   }
+]
+
+instance_group [
+  {
+    count: 1
+    kind : KIND_CPU
+  }
 ]
```
Comment (GuanLuo) [inline qa/L0_python_api/test_model_repository/identity/config.pbtxt:52]: new line

File: src/http_server.h [hunk 2]
```diff
@@ -134,6 +134,21 @@ class HTTPMetricsServer : public HTTPServer {
...
 
+  // Used by tritonfrontend python bindings
+  // Unifying Create function called from tritonfrontend.h
+  // TraceManager, RestrictedFeatures, and SharedMemoryManager objects
+  // are not propagated or used by HTTPMetricsServer
+  static TRITONSERVER_Error* Create(
+      std::shared_ptr<TRITONSERVER_Server>& server,
+      const UnorderedMapType& options,
+      triton::server::TraceManager*
+          trace_manager,  // Not used HTTPMetricsServer
```
Comment (GuanLuo) [inline src/http_server.h:145]: Don't do that just to make a "unified" interface..

File: src/python/tritonfrontend/_c/tritonfrontend_pybind.cc [hunk 3]
```diff
@@ -82,6 +82,15 @@ PYBIND11_MODULE(tritonfrontend_bindings, m)
                       triton::server::grpc::Server,
                       triton::server::grpc::Server>::StopService);
 #endif  // TRITON_ENABLE_GRPC
+
+#ifdef TRITON_ENABLE_METRICS
+  py::class_<TritonFrontend<HTTPServer, HTTPMetricsServer>>(
```
Comment (GuanLuo) [inline src/python/tritonfrontend/_c/tritonfrontend_pybind.cc:87]: following the previous comment, just specialize the template on metrics server

**Review by GuanLuo — DISMISSED** [review]

File: src/python/tritonfrontend/_api/_metrics.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    TritonFrontendMetrics,
+)
+
+
+class Metrics:
+    @dataclass
+    class Options:
+        address: str = "0.0.0.0"
+        port: int = Field(8002, ge=0, le=65535)
+        thread_count: int = Field(8, ge=0)
```
Comment (rmccorm4) [inline src/python/tritonfrontend/_api/_metrics.py:45]: Do we default to 8 threads in `main.cc` too?

Comment (KrishnanPrash) [reply src/python/tritonfrontend/_api/_metrics.py]: Thank you for the catch, changed it to 1.

File: src/http_server.h (hunk 2 above)
Comment (KrishnanPrash) [reply src/http_server.h]: Thanks for the suggestion! Removed unused args from `Create(...)` function.

File: docs/customization_guide/tritonfrontend.md [hunk 5]
```diff
@@ -57,14 +57,18 @@ Note: `model_path` may need to be edited depending on your setup.
 
 2. Now, to start up the respective services with `tritonfrontend`
 ```python
-from tritonfrontend import KServeHttp, KServeGrpc
+from tritonfrontend import KServeHttp, KServeGrpc, Metrics
```
Comment (rmccorm4) [inline docs/customization_guide/tritonfrontend.md:60]: I don't love that the `Metrics` object is a web server, so it makes me wonder if we should rename these down the line, ex: `KServeHttpService`, `MetricsService`, etc

But I don't have a strong opinion on an alternative right now so I think it's fine, just mentioning for later. We will probably be restructing some packaging and naming in the near-mid future.

File: qa/L0_python_api/test_model_repository/identity/config.pbtxt [hunk 6]
```diff
@@ -41,4 +41,10 @@ output [
     data_type: TYPE_STRING
     dims: [ 1 ]
   }
-]
\ No newline at end of file
+]
+instance_group [
+  {
+    count: 1
+    kind : KIND_CPU
```
Comment (rmccorm4) [inline qa/L0_python_api/test_model_repository/identity/config.pbtxt:48]: Did you ever investigate the gpu label thing?

File: qa/L0_python_api/testing_utils.py [hunk 7]
```diff
@@ -73,16 +75,31 @@ def teardown_service(service: Union[KServeHttp, KServeGrpc]) -> None:
...
+    Sets up a client to communicate with the Server through the respective protocol.
+    """
     return frontend_client.InferenceServerClient(url=url)
 
 
-def teardown_client(client) -> None:
+def teardown_client(
+    client: Union[
+        "tritonclient.http.InferenceServerClient",
+        "tritonclient.grpc.InferenceServerClient",
```
Comment (rmccorm4) [inline qa/L0_python_api/testing_utils.py:90]: I think this type hint is correct, and the other places you use `Union` is missing `InferenceServerClient`. You could probably also use `InferenceServerClientBase` though it'd be a bit less strict.

File: qa/L0_python_api/testing_utils.py [hunk 8]
```diff
@@ -73,16 +75,31 @@ def teardown_service(service: Union[KServeHttp, KServeGrpc]) -> None:
...
+    ]
+) -> None:
     client.close()
 
 
-# Sends an inference to test_model_repository/identity model and verifies input == output.
-def send_and_test_inference_identity(frontend_client, url: str) -> bool:
+def send_and_test_inference_identity(
+    frontend_client: Union["tritonclient.http", "tritonclient.grpc"], url: str
```
Comment (rmccorm4) [inline qa/L0_python_api/testing_utils.py:97]: See other comment on type hints, apply throughout

File: src/python/tritonfrontend/_api/__init__.py [hunk 9]
```diff
@@ -37,3 +37,10 @@
     # TRITON_ENABLE_GRPC=OFF
     # TritonFrontendGrpc Package was not present
     pass
+
+try:
+    from ._metrics import Metrics
+except ImportError:
+    # TRITON_ENABLE_Metrics=OFF
```
Comment (rmccorm4) [inline src/python/tritonfrontend/_api/__init__.py:44]: Make sure `L0_build_variants` passes

File: src/python/tritonfrontend/_api/_kservegrpc.py [hunk 10]
```diff
@@ -90,47 +88,38 @@ def __post_init__(self):
...
     def __enter__(self):
         self.triton_frontend.start()
         return self
 
+    @handle_triton_error
     def __exit__(self, exc_type, exc_value, traceback):
         self.triton_frontend.stop()
         if exc_type:
-            raise ERROR_MAPPING[exc_type](exc_value) from None
+            raise exc_type(exc_value) from None
```
Comment (rmccorm4) [inline src/python/tritonfrontend/_api/_kservegrpc.py:117]: Same comment: https://github.com/triton-inference-server/server/pull/7720#discussion_r1811533149

Why did this one keep the `from None` but the others didn't?

File: src/python/tritonfrontend/_api/_metrics.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    TritonFrontendMetrics,
+)
+
+
+class Metrics:
+    @dataclass
+    class Options:
+        address: str = "0.0.0.0"
+        port: int = Field(8002, ge=0, le=65535)
+        thread_count: int = Field(1, ge=0)
```
Comment (rmccorm4) [inline src/python/tritonfrontend/_api/_metrics.py:45]: Can thread count be zero?

File: qa/L0_python_api/test_model_repository/identity/config.pbtxt (hunk 6 above)
Comment (KrishnanPrash) [reply qa/L0_python_api/test_model_repository/identity/config.pbtxt:48]: Investigated a bit, but did not find the root cause. Will create a ticket in my backlog with hopefully a more consistent reproducer.

**Review by rmccorm4 — COMMENTED** [review]
Only minor comments - nice work!

File: src/python/tritonfrontend/_api/_metrics.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    TritonFrontendMetrics,
+)
+
+
+class Metrics:
+    @dataclass
+    class Options:
+        address: str = "0.0.0.0"
+        port: int = Field(8002, ge=0, le=65535)
+        thread_count: int = Field(1, ge=1)
```
Comment (rmccorm4) [inline src/python/tritonfrontend/_api/_metrics.py:45]: Please fix lower bound in the http/kserve classes too

Comment (GuanLuo) [reply src/python/tritonfrontend/_api/_metrics.py:45]: Should have added test case for 0 thread count, and probably for other numerical options

File: src/python/tritonfrontend/_api/_metrics.py (hunk 11 above)
Comment (KrishnanPrash) [reply src/python/tritonfrontend/_api/_metrics.py]: Change restriction for thread count to be `thread_count > 0`

File: src/python/tritonfrontend/_api/__init__.py (hunk 9 above)
Comment (KrishnanPrash) [reply src/python/tritonfrontend/_api/__init__.py:44]: Passing pipeline with `L0_build_variants`. The CI Pipeline ID is 19814908.

File: qa/L0_python_api/test_kserve.py [hunk 13]
```diff
@@ -271,6 +302,62 @@ def callback(user_data, result, error):
...
+        status_code, _ = utils.get_metrics(metrics_url)
+
+        assert status_code == 200
+
+        utils.teardown_service(service)
+        utils.teardown_server(server)
+
+    @pytest.mark.parametrize("frontend, url", [METRICS_ARGS])
+    def test_metrics_update(self, frontend, url):
+        # For this test
```
Comment (rmccorm4) [inline qa/L0_python_api/test_kserve.py:333]:
```suggestion
```

**Review by rmccorm4 — APPROVED** [review]

**Pull request merged**